File: src/pipeidea/sensitivity.py
"""Heuristics for detecting prompts that need a more grounded register."""

from __future__ import annotations

import re
from dataclasses import dataclass


_SENSITIVE_PATTERNS = (
    r"\bgaza\b",
    r"\bisrael\b",
    r"\bpalestin(?:e|ian|ians)\b",
    r"\bhamas\b",
    r"\bceasefire\b",
    r"\bwar\b",
    r"\bconflict\b",
    r"\bgenocide\b",
    r"\bethnic cleansing\b",
    r"\bterror(?:ism|ist|ists)?\b",
    r"\bhostage(?:s)?\b",
    r"\brefugee(?:s)?\b",
    r"\bdisplacement\b",
    r"\bfamine\b",
    r"\bhumanitarian\b",
    r"\batrocit(?:y|ies)\b",
    r"\babuse\b",
    r"\bsuicid(?:e|al)\b",
    r"\bself-harm\b",
    r"\boverdose\b",
    r"\bepidemic\b",
    r"\bpandemic\b",
    r"\boutbreak\b",
    r"\bmassacre\b",
    r"\bshooting\b",
)

_SENSITIVE_RE = re.compile("|".join(_SENSITIVE_PATTERNS), re.IGNORECASE)


@dataclass(frozen=True)
class SensitivityAssessment:
    """Structured result for prompt sensitivity checks."""

    is_sensitive: bool
    reason: str | None = None


def assess_prompt_sensitivity(seeds: list[str], mode: str) -> SensitivityAssessment:
    """Return whether a prompt likely needs grounded, high-stakes handling."""
    haystack = " ".join(seeds).strip()
    if not haystack:
        return SensitivityAssessment(is_sensitive=False)

    match = _SENSITIVE_RE.search(haystack)
    if not match:
        return SensitivityAssessment(is_sensitive=False)

    trigger = match.group(0).strip().lower()
    reason = (
        f"The user's prompt includes a live high-stakes topic ({trigger}). "
        f"Respond with humane, concrete, non-decorative language."
    )
    if mode == "collision":
        reason += " Keep the collision tightly anchored and avoid playful metaphor."
    else:
        reason += " Prefer direct mechanisms and grounded framing over bloom-style abstraction."
    return SensitivityAssessment(is_sensitive=True, reason=reason)

File: cloudflare/worker/src/sensitivity.mjs
const SENSITIVE_PATTERNS = [
  /\bgaza\b/i,
  /\bisrael\b/i,
  /\bpalestin(?:e|ian|ians)\b/i,
  /\bhamas\b/i,
  /\bceasefire\b/i,
  /\bwar\b/i,
  /\bconflict\b/i,
  /\bgenocide\b/i,
  /\bethnic cleansing\b/i,
  /\bterror(?:ism|ist|ists)?\b/i,
  /\bhostage(?:s)?\b/i,
  /\brefugee(?:s)?\b/i,
  /\bdisplacement\b/i,
  /\bfamine\b/i,
  /\bhumanitarian\b/i,
  /\batrocit(?:y|ies)\b/i,
  /\babuse\b/i,
  /\bsuicid(?:e|al)\b/i,
  /\bself-harm\b/i,
  /\boverdose\b/i,
  /\bepidemic\b/i,
  /\bpandemic\b/i,
  /\boutbreak\b/i,
  /\bmassacre\b/i,
  /\bshooting\b/i
];

export function assessPromptSensitivity(seeds, mode) {
  const haystack = String((seeds || []).join(" ")).trim();
  if (!haystack) {
    return { isSensitive: false, reason: null };
  }

  for (const pattern of SENSITIVE_PATTERNS) {
    const match = haystack.match(pattern);
    if (!match) {
      continue;
    }

    const trigger = String(match[0] || "").trim().toLowerCase();
    const base =
      `The user's prompt includes a live high-stakes topic (${trigger}). ` +
      "Respond with humane, concrete, non-decorative language.";
    const reason = mode === "collision"
      ? `${base} Keep the collision tightly anchored and avoid playful metaphor.`
      : `${base} Prefer direct mechanisms and grounded framing over bloom-style abstraction.`;
    return { isSensitive: true, reason };
  }

  return { isSensitive: false, reason: null };
}
